### PR TITLE
52 transaction feature dependent ops

### PR DESCRIPTION
**Pull request by keski**
This pull request addresses issue #52, allowing variables to be exported for use in dependent transactions.

### CODE REVIEW

**Comment by keski** [comment]
@epii-1, could you try the new feature out and see if you can break it? :)

**Comment by hartig** [comment]
Robin, please also include the corresponding entries to CHANGELOG.md into your PR.

**Comment by epii-1** [comment]
Robin, I'm trying to test this, but I'm getting stuck surprisingly early:

``` graphQL
mutation {
    createPlanet(data:{name: "Planet1"}) {
        id
    }
}```
works just fine. But 
``` graphQL
mutation {
    createPlanet(data:{name: "Planet1"}) {
        id @export(as: "planet1ID)
    }
}```
throws errors early and never enters our `create` function in the driver. I suspect that I am either using the wrong syntax, or that the `@export` directive is not working as expected.

Any quick pointers on what I am (probably) doing wrong?

**Comment by keski** [comment]
I think you need to declare the variable for the mutation:
```graphl
mutation($Planet1ID: String!) {
    createPlanet(data:{name: "Planet1"}) {
        id @export(as: "planet1ID)
    }
}```
Then you also need to use the variable somewhere (e.g. an additional createPlanet operation).

**Comment by epii-1** [comment]
Thanks Robin!
``` graphQL
mutation($name1: String!) {
                createPlanet(data:{name: "testname"}) {
                    name @export(as: "name1")
                }
                createSpecies(data:{name: $name1}) {
                    name
                }
            }
```
Worked just as expected when I got the syntax correct.
However,
``` graphQL
mutation($name1: String!) {
                createSpecies(data:{name: $name1}) {
                    name
                }
                createPlanet(data:{name: "testname"}) {
                    name @export(as: "name1")
                }
            }
```
did not work as expected, nor did it throw an error. The `data.createSpecies.name` value returned is simply `null`. I.e. the variabel is actually never used. Should it then not throw the same error as 
``` graphQL
mutation($name1: String!) {
                createSpecies(data:{name: "testnam2") {
                    name
                }
                createPlanet(data:{name: "testname"}) {
                    name @export(as: "name1")
                }
            }
```
does? (Over unused variable).

When tried with
``` graphQL
mutation($name1: String!) {
                createPlanet(data:{name: $name1}) {
                    name
                }
                createSpecies(data:{name: "testname"}) {
                    name @export(as: "name1")
                }
            }
```
if fails as `name` is a required field in `planets` (expected).

This only seem to happen with scalars fields thou. Trying the same thing with a type field and connect fails in the connect.

TLDR: Small inconsistency when trying to use scalar variables before they are assigned values.


Have you checked what happens with datetime fields by the way? As there might be a problem with the casting there. Or do we simply not allow exports of datetimefiels?

I'll see if I can find anything else.

**Comment by keski** [comment]
Nice catch. Indeed the case where the variable is referenced before being exported should fail. I'll fix it and get back to you. From Monday's meeting we decided to allow all scalar types, which should be fixed as soon as I add field vs variable validation #101.

**Comment by epii-1** [comment]
``` graphQL
mutation($ID1: ID!) {
                createSpecies(data:{name: "test"
                    }) {
                    id @export(as: "ID1")
                    name
                }
                updateSpecies(id: $ID1,
                            data: {name: "test2"}){
                    name
                }
            }
```
Fails with missing document key? I suspect the same things happens if I add deletions and try to delete it too.

**Comment by keski** [comment]
This should now have been resolved. @epii-1 , could you verify that it works on your end?

**Comment by epii-1** [comment]
Robin, I still get the same behavior with 
```
mutation($name1: String!) {
                createSpecies(data:{name: $name1}) {
                    name
                }
                createPlanet(data:{name: "testname"}) {
                    name @export(as: "name1")
                }
            }
```
passing, but the name for the newly created species being `null`?

I also double checked that dateTime fields works as expected with export, and indeed they do work as expected (no weird casting).

**Comment by keski** [comment]
Okay, I know what's wrong. Will push fix later today.

**Comment by keski** [comment]
I have fixed the bug but cannot not re-build `LiUGraphQL/graphql-js` due to some recent changes in the GraphQL dependencies. Looking into this now.

**Comment by keski** [comment]
I have now merged the latest changes in `graphql/graphql-js` into `LiUGraphQL/graphql-js`. Below follow a subset of the mutations I've used for testing.

These should fail:
```graphql
mutation TypeMismatch1 ($name:String!) {
  createHuman(data:{
    name: "Han Solo"
    appearsIn: EMPIRE
  }){
    id @export(as: "name")
    name 
  },
  createDroid(data:{
    name: $name
    appearsIn: EMPIRE
  }){
    id
    name 
  }
}

mutation TypeMismatch2 ($episodes:Episode) {
  createHuman(data:{
    name: "Han Solo"
    appearsIn: EMPIRE
  }){
    id
    name
    appearsIn @export(as: "name")
  },
  createDroid(data:{
    name: "C3PO"
    appearsIn: $episodes
  }){
    id
    name 
  }
}

mutation ExportTwice ($episodes:[Episode]!) {
  createHuman(data:{
    name: "Han Solo"
    appearsIn: EMPIRE
  }){
    id
    name
    appearsIn @export(as: "episodes")
  },
  createDroid(data:{
    name: "C3PO"
    appearsIn: $episodes
  }){
    id
    name 
    appearsIn @export(as: "episodes")
  }
}

mutation ReferenceBeforeAssignment ($name:String!) {
  createHuman(data:{
    name: $name
    appearsIn: EMPIRE
  }){
    id
    name @export(as:"name")
  }
}
 
mutation ExportUndeclared {
  createHuman(data:{
    name: "Han Solo"
    appearsIn: EMPIRE
  }){
    id
    name @export(as: "name")
  }
}

mutation ExportWithDefaultValue($primaryFunction:String="")  {
  m1:createDroid(data:{
    name: "C3PO" 
    appearsIn: EMPIRE
    primaryFunction: "Complain"
  }){
    id
    name
    primaryFunction @export(as:"primaryFunction")
  }
  m2:createDroid(data:{
    name: "C3PO_v2" 
    appearsIn: EMPIRE
    primaryFunction: $primaryFunction
  }){
    id
    name
    primaryFunction
  }
}

mutation ExportWithPassedValue($primaryFunction:String)  {
  m1:createDroid(data:{
    name: "C3PO" 
    appearsIn: EMPIRE
    primaryFunction: "Complain"
  }){
    id
    name
    primaryFunction @export(as:"primaryFunction")
  }
  m2:createDroid(data:{
    name: "C3PO_v2" 
    appearsIn: EMPIRE
    primaryFunction: $primaryFunction
  }){
    id
    name
    primaryFunction
  }
}
# With query variables: {"primaryFunction":""}
```
While the following should be accepted:
```graphql
mutation HumanWithTwoDroidFriends(
  $humanID: ID!
  $droid1ID: ID!
  $droid2ID: ID!
  $episodes: [Episode]!
) {
  m1:createHuman(data: { name: "Luke", appearsIn: [EMPIRE, JEDI, NEWHOPE] }) {
    id @export(as: "humanID")
    appearsIn @export(as:"episodes")
  }
  m2:createDroid(data: { name:"C3PO", appearsIn: $episodes }){
    id @export(as: "droid1ID")
  }
  m3:createDroid(data: { name: "R2D2", appearsIn: $episodes }){
    id @export(as: "droid2ID")
  }
  m4:createFriendsEdgeFromHuman(data: {
    sourceID: $humanID
    targetID: $droid1ID
  }){
    id
    target {
      id
    }
    source {
      id
    }
  }
  m5:createFriendsEdgeFromHuman(data: {
    sourceID: $humanID
    targetID: $droid2ID
  }){
    id
    target {
      id
    }
    source {
      id
    }
  }
}
```

**Comment by epii-1** [comment]
Looks good and is working fine for the tests I could think of.
